### DISCUSSION TITLE
Scheduled builds of non-default branches

### DISCUSSION THREAD

**Discussion by mkurz** [community/community#16107] [Other Feature Feedback, Questions, & Ideas]
Right now it's just not possible to run scheduled builds for non-default branches.
I am **not** talking about checking out another branch from a workflow defined in the default branch:

```yaml
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: my_branch
````
Even if you check out the branch, you still run the "current" workflow of the default branch where that checkout step is defined.

There is [a discussions going on](https://github.community/t/scheduled-builds-of-non-default-branch/16306) and I think [this comment](https://github.community/t/scheduled-builds-of-non-default-branch/16306/15) describes very well why checkout alone won't help.

The problem for us is that we have a workflow defined in a non-default branch that runs on `pull_request` and `push` events, which is working fine, however we also want to run the same tests for that branch scheduled, each night as well. This way we can run the same tests + additional, more thorough, tests, which take much longer, each night (by conditionally checking `if: github.event_name == 'schedule'`).

Right now the only, very dirty, workaround is to copy a workflow defintion file from a non-default branch to the default branch and use the workaround to check out non-default branch like shown in the snippet above.

Maintaining two exact same copies of the same workflow is very cumbersome as you can imagine and its very likely they become out of sync at some point. 

I can imagine and understand that there are technical limitations of checking every branch in every repo for scheduled workflows, however if you give repo admins a setting page where they can define additional branches, besides the default branch, where GitHub should look for scheduled workflows that should be enough. Instead of a settings page you could also offer a new workflow syntax. E.g. something like `.github/workflows/scheduled.yml`:
```yaml
on:
  schedule:
    cron:
      - 14 18 * * *
workflows:
- my_workflow.yml@my_branch
```
(Just an idea 😉)

Thanks!

**Comment by jsoref** [comment]
One approach would be to use [nektos/act](http://github.com/nektos/act) -- have the schedule check out the branch and then tell it which workflow(s) to run.

Alternatively, it's possible that you can do what you want with a reusable workflow:
https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow

**Reply by mkurz** [reply]
Using a reusable workflow is not possible for us because the workflow in the non-default branch itself already calls multiple reusable workflows. And "_Reusable workflows can't call other reusable workflows._": https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations

**Reply by jsoref** [reply]
Welp that rules out one approach.

otoh, nektos/act supports reusable workflows.

**Reply by mkurz** [reply]
ok, however using nektos/act still is a workaround and IMHO there should be a nicer, official way to achieve what we want.

**Reply by jsoref** [reply]
I'd like a pony 🐴 . (I'll settle for a unicorn 🦄  I can ride w/o having to shelter 🐴 🏚️ .)

I also would hate for a repository to suddenly trigger scheduled tasks on a thousand branches when I fork it. I also don't look forward to scheduled tasks that error ❌  because when I fork them I delete their non default branches.

Please do me a favor and set your scheduled workflow to not run (unless some secret is set) that way on average forks won't trip on it.

**Reply by mkurz** [reply]
@jsoref Look here please: https://github.com/github/feedback/discussions/16109

**Reply by mkurz** [reply]
I also don't look forward to even trigger the main branch's scheduled workflow for forks.

**Reply by mkurz** [reply]
Even better @jsoref: https://github.com/github/feedback/discussions/12180

**Reply by GordonSmith** [reply]
> Using a reusable workflow is not possible for us because the workflow in the non-default branch itself already calls multiple reusable workflows. And "_Reusable workflows can't call other reusable workflows._": https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations

FYI These limitations have been improved to "4 levels of reusable workflows"

**Comment by skorobogatydmitry** [comment]
Same issue, a bit different use-case.
Environment protection rules doesn't allow to start jobs from the default branch. So, despite that I specify branch explicitly in the scheduled workflow, GH doesn't allow it to execute. Same happens with the reusable workflow, as the called inherits branch from the caller.

**Comment by Pr0methean** [comment]
I need this feature too: I'm trying to compare two GCs for a JVM program, https://github.com/Pr0methean/OcHd-KotlinBuild/commit/db17068d0bc6e3b85416bd720dfde1735a99ed1d (Shenandoah) and https://github.com/Pr0methean/OcHd-KotlinBuild/commit/65871e5f89833d997879d31357b7ed87b00ae67c (ZGC), and launching them at different times of day or days of the week could easily bias the results. The only alternative solution I can come up with involves willingly regressing mainline for several days while I learn which commit executes its CI/CD faster.

**Comment by jbrsb** [comment]


**Reply by mkurz** [reply]
Sorry, no solution so far. We just don't run scheduled jobs for non default branches now...

**Comment by haroldo-bonette** [comment]
also struggling to move a very simple pipeline from Jenkins to GHA...
we need to trigger on schedule jobs on predefined branches, we can achieve that with a reusable workflow and defining the ref on checkout, but if you define environments and env protection it wont work as the Default branch will propagate to the reusable workflow:
it would really be great if we could define branches on schedule trigger:

something like:
``` yml
on:
  schedule:
    - cron: '*/5 * * * *'
      branches:
        - Dev
        - Int
```

**Reply by thetredev** [reply]
Same here. I have a separate branch for creating a container image which I would like to run on a schedule. The default branch then uses that container image to do other things. This syntax would solve all my issues!

**Reply by mkurz** [reply]
I also like this syntax, but I would let the user even define _which yml file_ in which branch to use. Because the given example only defines the branches, but how to define which yml file from the branch to actually use? Most likely the one with the same file name, however with following syntax that is even better and more flexible IMHO:

```yml
on:
  schedule:
    - cron: '*/5 * * * *'
      workflows:
        - my_workflow.yml@some_branch
        - my_other_workflow.yml@some_other_branch
```

**Comment by thetredev** [comment]
GitHub is currently asking for feedback on Actions (just click on the Actions tab on any of your projects). I just gave them feedback about this topic. Hope it has an effect.

**Comment by VitaliMorgun** [comment]
Is there any solution so far to define branches on schedule trigger as we do it on push trigger for example?

**Reply by jsoref** [reply]
Write a workflow on the default branch with a schedule, and have it use [`workflow_call`](https://docs.github.com/en/actions/using-workflows/reusing-workflows) to call your workflow on the other branch

**Reply by mco576** [reply]
This doesn't work when using environments protected by branch rules

**Comment by mikeedwards-rs** [comment]
Just adding a +1 for this feature. Surprised it isn't already there.

**Comment by EwoutH** [comment]
Yes this would be very useful, especially if you have maintenance branches.

**Comment by luis-m-gonzalez** [comment]
I need this too

**Comment by snowdream** [comment]
I need this too

**Comment by bencat-sixense** [comment]
Same same :)

**Comment by rbankole** [comment]
plus 1 for this feature

**Comment by coralkashri** [comment]
+1.
Sometimes workflows doesn't allow to update a different branch than the current one, and having a scheduled workflow on another branch would be helpful.

**Comment by vc1492a** [comment]
+1 
We would like to use github actions for nightly builds and deploys of our `dev` branch.

**Reply by jsoref** [reply]
- https://github.com/community/community/discussions/16107#discussioncomment-8786272

**Comment by tiesteaz** [comment]
I'm migrating my pipeline from Gitlab to Github, and knowing that in Gitlab you are specifying branch, environment variables, enabled/disabled and even Cron time zone, all of this on the same UI form, it feels so easy in Gitlab compared to what you have to do in Github...

And then you can click "Play button" to run it manually

I wish there were product managers in Github who review other products to understand what customers may struggling for

![image](https://github.com/user-attachments/assets/a526e0ec-70fc-454d-a988-1d5f0ac99d12)
![image](https://github.com/user-attachments/assets/932b836f-f65f-45ad-864d-27705c01e18b)

**Reply by damjabra-flutterint** [reply]
yeah gitlab was a breeze compared to gh.. not just pipelines, we are having trouble with merge conflicts where there are none, and comparing current and incoming files, where you can actually see the difference, gh is not reporting as modified file.. freaky

**Reply by Quanalogy** [reply]
Yea having the same feeling moving from a company where we've used GitLab for 5 years and then moving to a new company where we use GitHub. It really does feel like a downgrade on many points.

**Reply by dimaBondarenko** [reply]
Gosh, I miss this Gitlab feature so much. Now with Github it feels like I've switched from a supercar to a horse carriage 😭

**Comment by cscheffauer** [comment]
+1

**Comment by LanderV-Vintecc** [comment]
+1

**Comment by anastasiyamazheika** [comment]
+1

**Comment by grilla99** [comment]
+1

**Comment by mertkucukkuru-aily** [comment]
Seems like no-one implemented this solution?

**Comment by arnabonline** [comment]
This would be useful
